### PR TITLE
Feat: add streaming support to openai crate

### PR DESCRIPTION
**Pull request by DMJ16**
~This is a work in progress for adding streaming support to the openai crate.~

~I believe it should poll and resolve the stream in this case:~
~`let message_text = res.primary_textual_output().await.unwrap();`~

~But also expose the ability to pass the stream on in the case of a web api sending the stream to a client. Added to the openai crate's Output an `as_stream` method for this.~

Basic streaming implementation for openai. Unless the streaming step is the final step or a one of one run, the stream will be consumed.

### CODE REVIEW

File: llm-chain-openai/src/chatgpt/step.rs [hunk 1]
```diff
@@ -50,25 +50,38 @@ impl ToString for Model {
...
         let prompt = prompt.into();
-        Step { model, prompt }
+        Step {
+            model,
+            prompt,
+            stream,
+        }
     }
-    pub fn for_prompt_and_model<P: prompt::Prompt>(model: Model, prompt: P) -> Step {
+    pub fn for_prompt_and_model<P: prompt::Prompt>(
```
Comment (williamhogman) [inline llm-chain-openai/src/chatgpt/step.rs:72]: Maybe we can make it `$methodname_with_streaming` for the streaming methods. We will switch a builder pattern soon

File: llm-chain-openai/src/chatgpt/output.rs [hunk 2]
```diff
@@ -11,54 +76,138 @@
...
+                .iter()
+                .map(|choice| choice.message.content.clone())
+                .collect(),
+            OutputInner::Stream(stream) => stream.primary_textual_output_choices().await,
+        }
     }
 }
 
-/// Implement the Deref trait to provide access to the underlying CreateChatCompletionResponse.
+/// Implement the Deref trait to provide access to the underlying OutputInner.
```
Comment (williamhogman) [inline llm-chain-openai/src/chatgpt/output.rs:186]: I suppose we could drop deref at this point

**Review by williamhogman — APPROVED** [review]
This is really starting to come together nicely! Well done

**Comment by DMJ16** [comment]
Made a first round of updates. Builder pattern sounds great. Is there something I can do to get this to a prod-ready state? Maybe example stream usage? What are the test requirements for now?

**Comment by williamhogman** [comment]
Hey,

I think an example should do for now.  For the example we could try a two step sequential chain, where the first step is executed normally and the second one is streamed :)

**Comment by DMJ16** [comment]
I've got streaming working for primary_textual_output as well as passing the stream through to a client with sse. However, I cannot get it working when calling to_chain on the streaming step. I'm losing the data somewhere.

**Comment by DMJ16** [comment]
@williamhogman I did some hw and tried a bunch of things for the stream cloning/buffer etc as we discussed, but I cannot come up with anything that works. Latest code circumvents the issue of cloning stream/peeking/ making pinned box stream sync etc.  

- It ignores polling the stream and getting the choices/output if and only if the final step is a stream. Not ideal but curious what you think.  
- Added an example which just modifies the other chain, so you can see the explicit choices call and also commented out how to access the stream for web apps etc.

**Comment by williamhogman** [comment]
> @williamhogman I did some hw and tried a bunch of things for the stream cloning/buffer etc as we discussed, but I cannot come up with anything that works. Latest code circumvents the issue of cloning stream/peeking/ making pinned box stream sync etc.
> 
> * It ignores polling the stream and getting the choices/output if and only if the final step is a stream. Not ideal but curious what you think.
> * Added an example which just modifies the other example of using seq chain, so you can see the explicit choices method call and also I commented beneath the final println with how to access the stream for web apps etc.

I think this approach is worthwhile. Certainly output streaming is a big deal so even if we weren't able to figure out how to get it to behave 100% we get the right result for the end user. 

The Step changes for mainline just made it in so you'll have to rebase but this looks really good 🎉

**Pull request merged**